src/hilok.cpp

```diff
@@ -1 +1,273 @@
+
+        ++it_to;
+        if (common) {
+            // common ancestor nodes can be ignored
+            from_key = {cur_to, *it_from};
+            ++it_from;
+            if (key == from_key) {
+#ifdef HILOK_TRACE
+                std::cout << "ig: " << key.first << "/" << key.second << std::endl;
+#endif
+                auto it = m_map.find(key);
+                cur_to = it->second;
+                cur_from = it->second;
+                continue;
+            }
+            common = false;
+        }
+
+        if (it_to != it_to.end()) {
+            // non-leaf destination
+            // get or create node, as needed for the destination
+            // clone lock counts && thread ids from the leaf
+
+            auto it = m_map.find(key);
+            if (it == m_map.end()) {
+                cur_to = m_map[key] = std::make_shared<HiKeyNode>(key, m_recursive);
+            } else {
+                cur_to = it->second;
+            }
+
+#ifdef HILOK_TRACE
+            std::cout << "clon lk: " << key.first << "/" << key.second << ":" << cur_to << " " << leaf_from_node->m_mut.m_ex_cnt << std::endl;
+#endif
+            // copy lock counts from the leaf to the ancestor
+            if (!cur_to->m_mut.unsafe_clone_lock_shared(leaf_from_node->m_mut, block, secs)) {
+                throw HiErr("unable to lock rename dest");
+            }
+        } else {
+            cur_to.reset(); // refcount for erase
+
+            while (it_from != it_from.end()) {
+                // uncommon ancestor of source must be released
+                ++it_from;
+
+                auto it = m_map.find(from_key);
+
+                // we already tested for this above, and we have a mutex, should never happen
+                assert(it != m_map.end());
+
+                cur_from = it->second;
+
+#ifdef HILOK_TRACE
+                std::cout << "clon un: " << from_key.first << "/" << from_key.second << ":" << cur_from << std::endl;
+#endif
+                // unlock uncommon ancestors of the source
+                cur_from->m_mut.unsafe_clone_unlock_shared(leaf_from_node->m_mut);
+
+                from_key.first.reset(); // refcount for erase
+                // could have went to 0
+                erase_unsafe(cur_from);
+            
+                from_key = {cur_from, *it_from};
+            }
+
+            // keep leaf locks, change key
+            m_map.erase(leaf_from_node->m_key);
+            leaf_from_node->m_key = key;
+            m_map[key] = leaf_from_node;        
+       } 
+    }
+
+
+}
+
+
+void HiLok::erase_safe(std::shared_ptr<HiKeyNode> &ref) {
+    std::lock_guard<std::mutex> guard(m_mutex);
+    erase_unsafe(ref);
+}
+
+void HiLok::erase_unsafe(std::shared_ptr<HiKeyNode> &ref) {
+    // lazy speedup, there can be lots of refs (parent->child ref, caller refs, etc.)
+    // but if there are too many, we know the exclusive cannot work and is not worth even trying
+    if (ref.use_count() <= 6) {
```

Comment:
If we get this nuber wrong (e.g. in a future version, there can be 7 refs with only one caller in some circumstances) we can end up not removing from the map, right? Assuming this would be caught by existing tests?

Reply:
yes, tests would catch cuz they all assert size is 0 at the end